File: sdk/dart/lib/src/blockchain/chat_chain_client.dart
/// Chat Chain client for Dart SDK
/// Handles identity, messaging, channels, governance on chat chain

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'client.dart';
import 'transaction.dart';

/// Chat Chain transaction types specific to messaging and identity
enum ChatChainTxType {
  registerUser,
  sendDirectMessage,
  createChannel,
  postToChannel,
  updateReputation,
  governance,
}

/// Chat Chain client for identity and messaging operations
class ChatChainClient extends BlockchainClient {
  /// Create new chat chain client
  ChatChainClient({
    String rpcUrl = 'http://localhost:8545',
    String? wsUrl,
  }) : super(
    config: BlockchainConfig(
      rpcUrl: rpcUrl,
      wsUrl: wsUrl,
    ),
  );

  /// Register user identity on chat chain
  @override
  Future<String> registerUser({
    required String userId,
    required String username,
    required String publicKey,
  }) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/chat/register_user'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'user_id': userId,
        'username': username,
        'public_key': publicKey,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to register user: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['tx_id'] as String;
  }

  /// Send direct message transaction
  @override
  Future<String> sendDirectMessage({
    required String messageId,
    required String senderId,
    required String recipientId,
    required String contentHash,
    required int payloadSize,
    String? relayNodeId,
  }) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/chat/send_message'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'message_id': messageId,
        'sender_id': senderId,
        'recipient_id': recipientId,
        'content_hash': contentHash,
        'payload_size': payloadSize,
        'relay_node_id': relayNodeId,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to send message: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['tx_id'] as String;
  }

  /// Create channel on chat chain
  @override
  Future<String> createChannel({
    required String channelId,
    required String creatorId,
    required String name,
    required String description,
    String? tokenRequirement,
    ChannelVisibility visibility = ChannelVisibility.public,
  }) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/chat/create_channel'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'channel_id': channelId,
        'creator_id': creatorId,
        'name': name,
        'description': description,
        'token_requirement': tokenRequirement,
        'visibility': visibility.toString().split('.').last,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to create channel: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['tx_id'] as String;
  }

  /// Post message to channel on chat chain
  @override
  Future<String> postToChannel({
    required String channelId,
    required String contentHash,
    required String messageId,
    required int payloadSize,
    required String senderId,
  }) async {
    final response = await http.post(
      Uri.parse('${config.rpcUrl}/chat/post_message'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'channel_id': channelId,
        'sender_id': senderId,
        'message_id': messageId,
        'content_hash': contentHash,
        'payload_size': payloadSize,
        'timestamp': DateTime.now().millisecondsSinceEpoch ~/ 1000,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to post message: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['tx_id'] as String;
  }

  /// Get user reputation
  Future<int> getReputation(String userId) async {
    final response = await http.get(
      Uri.parse('${config.rpcUrl}/chat/reputation/$userId'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to get reputation: ${response.body}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['reputation'] as int;
  }

  /// Get user transaction history
  Future<List<Map<String, dynamic>>> getUserTransactions(String userId) async {
    final response = await http.get(
      Uri.parse('${config.rpcUrl}/chat/transactions/$userId'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to get transactions: ${response.body}');
    }

    final data = jsonDecode(response.body) as List<dynamic>;
    return data.cast<Map<String, dynamic>>();
  }
}

File: sdk/typescript/src/blockchain/ChatChainClient.ts
// Chat Chain client for TypeScript SDK
// Handles identity, messaging, channels, governance on chat chain

import axios, { AxiosInstance } from 'axios';

export enum ChatChainTxType {
  RegisterUser = 'register_user',
  SendDirectMessage = 'send_direct_message',
  CreateChannel = 'create_channel',
  PostToChannel = 'post_to_channel',
  UpdateReputation = 'update_reputation',
  Governance = 'governance',
}

export interface ChatChainTransaction {
  id: string;
  tx_type: ChatChainTxType;
  sender: string;
  data: Record<string, any>;
  status: 'pending' | 'confirmed' | 'failed';
  confirmations: number;
  block_height: number;
  created_at: number;
}

export class ChatChainClient {
  private client: AxiosInstance;

  constructor(rpcUrl: string = 'http://localhost:8545') {
    this.client = axios.create({
      baseURL: rpcUrl,
      headers: { 'Content-Type': 'application/json' },
    });
  }

  async registerUser(userId: string, publicKey: Buffer): Promise<ChatChainTransaction> {
    const response = await this.client.post('/chat/register_user', {
      user_id: userId,
      public_key: publicKey.toString('base64'),
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async sendDirectMessage(
    sender: string,
    recipient: string,
    messageId: string,
  ): Promise<ChatChainTransaction> {
    const response = await this.client.post('/chat/send_message', {
      sender,
      recipient,
      message_id: messageId,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async createChannel(owner: string, channelId: string, name: string): Promise<ChatChainTransaction> {
    const response = await this.client.post('/chat/create_channel', {
      owner,
      channel_id: channelId,
      name,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async postToChannel(sender: string, channelId: string, messageId: string): Promise<ChatChainTransaction> {
    const response = await this.client.post('/chat/post_message', {
      sender,
      channel_id: channelId,
      message_id: messageId,
      timestamp: Math.floor(Date.now() / 1000),
    });
    return response.data;
  }

  async getReputation(userId: string): Promise<number> {
    const response = await this.client.get(`/chat/reputation/${userId}`);
    return response.data.reputation;
  }

  async getUserTransactions(userId: string): Promise<ChatChainTransaction[]> {
    const response = await this.client.get(`/chat/transactions/${userId}`);
    return response.data;
  }

  async getTransaction(txId: string): Promise<ChatChainTransaction | null> {
    try {
      const response = await this.client.get(`/chat/transaction/${txId}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async waitForConfirmation(txId: string, confirmations: number = 6, maxWaitMs: number = 30000): Promise<ChatChainTransaction> {
    const startTime = Date.now();
    while (Date.now() - startTime < maxWaitMs) {
      const tx = await this.getTransaction(txId);
      if (tx && tx.confirmations >= confirmations) {
        return tx;
      }
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
    throw new Error(`Transaction ${txId} did not confirm within ${maxWaitMs}ms`);
  }
}

export default ChatChainClient;
